### PR TITLE
Abstract the core module (Closes #97)

### LINKED ISSUE #97: Abstract the core module
There is a good chance that we could pull out all of Elasticsearch related codes from the `core` module and have it as abstract. After, we could move the Elasticsearch codes to its module and open up some room for other possible sinks for this project in future. 

What do you folks think?

### PR DESCRIPTION
**Pull request by szareiangm**
This pull request is focused on having anything related to Elasticsearch layer in `elasticsearch` module.  It is mostly moving things around to re-organize, i.e.:  Like renaming `ElasticesarchEmitter` to `Emitter` or `KinesisElasticsearchPipeline` to `KinesisPipeline`.

Two side-notes: 
- There is a `partition` notion in JsonRecord that will be used in the next pull request that I will make, but I kept it here just as a placeholder. I just `index` name as it value since all of the data goes to one index now.
- Also, `document_type` in Elastic search is getting deprecated, but I kept it in `elasticsearch` module.

CC: @Nafid

### CODE REVIEW

**Comment by snowplowcla** [comment]
@szareiangm has signed the Software Grant and Corporate Contributor License Agreement

File: core/src/main/scala/com.snowplowanalytics.stream/loader/Emitter.scala [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+               bufferByteLimit: Long
+             ) extends IEmitter[EmitterJsonInput] {
+
+
+
+
+  @throws[IOException]
+  def emit(buffer: UnmodifiableBuffer[EmitterJsonInput]): List[EmitterJsonInput] = {
+    attemptEmit(buffer.getRecords)
+  }
```
Comment (BenFradet) [inline core/src/main/scala/com.snowplowanalytics.stream/loader/Emitter.scala]: nit: we can remove the brackets

File: core/src/main/scala/com.snowplowanalytics.stream/loader/Emitter.scala [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+        case Some(s) => {
+          validRecords.foreach(recordTuple =>
+            recordTuple.map(record => record.map(r => s.store(r.json.toString, None, true))))
+          Nil
+        }
+        case None if validRecords.isEmpty => Nil
+        case _ => emit(validRecords)
+      }
+
+      (invalidRecords ++ rejects).asJava
```
Comment (BenFradet) [inline core/src/main/scala/com.snowplowanalytics.stream/loader/Emitter.scala]: why convert to java's list?

File: core/src/main/scala/com.snowplowanalytics.stream/loader/Emitter.scala [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+import Scalaz._
+
+// Scala
+import scala.collection.mutable.ListBuffer
+
+import scala.collection.immutable.{List => SList}
+import scala.collection.JavaConverters._
+
+
+import sinks.ISink
```
Comment (BenFradet) [inline core/src/main/scala/com.snowplowanalytics.stream/loader/Emitter.scala]: imports need to be sorted, you can look at the existing code for guidelines

File: core/src/main/scala/com.snowplowanalytics.stream/loader/Emitter.scala [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+
+  /**
+    * Emits good records to Elasticsearch and bad records to Kinesis.
+    * All valid records in the buffer get sent to Elasticsearch in a bulk request.
+    * All invalid requests and all requests which failed transformation get sent to Kinesis.
+    *
+    * @param records List of records to send to Elasticsearch
+    * @return List of inputs which Elasticsearch rejected
+    */
+  def emit(records: SList[EmitterJsonInput]): SMBuffer[EmitterJsonInput] = {
```
Comment (BenFradet) [inline core/src/main/scala/com.snowplowanalytics.stream/loader/Emitter.scala]: we shouldn't expose mutable data structures

File: core/src/main/scala/com.snowplowanalytics.stream/loader/StreamLoaderApp.scala [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+      case Left(e) =>
+        System.err.println(s"configuration error: $e")
+        System.exit(1)
+        None
+       case Right(c) => Some(c)
+    }
+
+    streamLoaderConf
+  }
+  val config:StreamLoaderConfig = parseConfig().get
```
Comment (BenFradet) [inline core/src/main/scala/com.snowplowanalytics.stream/loader/StreamLoaderApp.scala]: please don't use `get` on options

File: core/src/main/scala/com.snowplowanalytics.stream/loader/StreamLoaderApp.scala [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+
+  val tracker = config.monitoring.map(e => SnowplowTracking.initializeTracker(e.snowplow))
+
+    val goodSink = config.sink.good match {
+      case "stdout" => Some(new StdouterrSink)
+      case "elasticsearch" => None
```
Comment (BenFradet) [inline core/src/main/scala/com.snowplowanalytics.stream/loader/StreamLoaderApp.scala]: ?

File: core/src/main/scala/com.snowplowanalytics.stream/loader/Emitter.scala (hunk 1 above)
Comment (szareiangm) [reply core/src/main/scala/com.snowplowanalytics.stream/loader/Emitter.scala]: addressed

File: core/src/main/scala/com.snowplowanalytics.stream/loader/Emitter.scala (hunk 2 above)
Comment (szareiangm) [reply core/src/main/scala/com.snowplowanalytics.stream/loader/Emitter.scala]: `emit` needs Java list

File: core/src/main/scala/com.snowplowanalytics.stream/loader/Emitter.scala (hunk 3 above)
Comment (szareiangm) [reply core/src/main/scala/com.snowplowanalytics.stream/loader/Emitter.scala]: addressed

File: core/src/main/scala/com.snowplowanalytics.stream/loader/Emitter.scala (hunk 4 above)
Comment (szareiangm) [reply core/src/main/scala/com.snowplowanalytics.stream/loader/Emitter.scala]: `private def emit` solves it. right?

File: core/src/main/scala/com.snowplowanalytics.stream/loader/StreamLoaderApp.scala (hunk 6 above)
Comment (szareiangm) [reply core/src/main/scala/com.snowplowanalytics.stream/loader/StreamLoaderApp.scala]: it used to be like this from before:  https://github.com/snowplow/snowplow-elasticsearch-loader/blob/7861698a6da7fb309b0b7f4c7b51266d1b93a1c3/core/src/main/scala/com.snowplowanalytics.stream/loader/ElasticsearchSinkApp.scala#L150

Because the default sink is `elasticsearch` and goodSink is the extra addition.

File: core/src/main/scala/com.snowplowanalytics.stream/loader/StreamLoaderApp.scala (hunk 5 above)
Comment (szareiangm) [reply core/src/main/scala/com.snowplowanalytics.stream/loader/StreamLoaderApp.scala]: But it used to be the same as I did but in the child class: https://github.com/snowplow/snowplow-elasticsearch-loader/blob/7861698a6da7fb309b0b7f4c7b51266d1b93a1c3/elasticsearch/src/main/scala/com/snowplowanalytics/stream/loader/ElasticsearchHTTPSinkApp.scala#L28

I just moved it to have the variable in the trait. What is your suggestion?

File: core/src/main/scala/com.snowplowanalytics.stream/loader/Emitter.scala (hunk 4 above)
Comment (szareiangm) [reply core/src/main/scala/com.snowplowanalytics.stream/loader/Emitter.scala]: solved. This method is not exposed. Instead, emit is exposed and that one is good: `emit(buffer: UnmodifiableBuffer[EmitterJsonInput]): List[EmitterJsonInput]`

File: core/src/main/scala/com.snowplowanalytics.stream/loader/Emitter.scala (hunk 2 above)
Comment (szareiangm) [reply core/src/main/scala/com.snowplowanalytics.stream/loader/Emitter.scala]: `emit` needs Java type for its return value and this class mainly implements KCL emitter for Java.

**Comment by szareiangm** [comment]
@BenFradet I think I addressed all of your actionable comments. Please go for the next round of review if needed.

File: core/src/main/scala/com.snowplowanalytics.stream/loader/Emitter.scala [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+import scala.collection.immutable.{List => SList}
+import scala.collection.mutable.{Buffer => SMBuffer}
+import scala.collection.JavaConverters._
+
+// This project
+import sinks.ISink
+import clients.BulkSender
+
+/**
+  * ElasticsearchEmitter class for any sort of BulkSender Extension
```
Comment (BenFradet) [inline core/src/main/scala/com.snowplowanalytics.stream/loader/Emitter.scala]: wrong doc

File: core/src/main/scala/com.snowplowanalytics.stream/loader/Emitter.scala (hunk 2 above)
Comment (BenFradet) [reply core/src/main/scala/com.snowplowanalytics.stream/loader/Emitter.scala]: ah, can you add `override` to the methods which are from `IEmitter` that will make things clearer

**Comment by szareiangm** [comment]
@BenFradet I have no more comments to be addressed. Any chance for this PR to be merged?

File: core/src/main/scala/com.snowplowanalytics.stream/loader/Emitter.scala [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+
+  /**
+    * Emits good records to Sink and bad records to Kinesis.
+    * All valid records in the buffer get sent to the sink in a bulk request.
+    * All invalid requests and all requests which failed transformation get sent to Kinesis.
+    *
+    * @param records List of records to send
+    * @return List of inputs which the sink rejected
+    */
+  def emit(records: SList[EmitterJsonInput]): SMBuffer[EmitterJsonInput] = {
```
Comment (BenFradet) [inline core/src/main/scala/com.snowplowanalytics.stream/loader/Emitter.scala]: can we not have a buffer here?

File: core/src/main/scala/com.snowplowanalytics.stream/loader/Emitter.scala [hunk 9]
```diff
@@ -0,0 +1,179 @@
...
+  def shutdown(): Unit = bulkSender.close
+
+  /**
+    * Handles records rejected by the JsonTransformer or by Sink
+    *
+    * @param records List of failed records
+    */
+   def fail(records: List[EmitterJsonInput]): Unit = {
+    records.asScala.foreach {
+      _ match {
```
Comment (BenFradet) [inline core/src/main/scala/com.snowplowanalytics.stream/loader/Emitter.scala]: no need for that

File: core/src/main/scala/com.snowplowanalytics.stream/loader/StreamLoaderApp.scala [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+ */
+trait StreamLoaderApp extends App {
+
+  val executor: Validation[String, Runnable]
+  lazy val arguments = args
+
+  private def parseConfig(): Option[StreamLoaderConfig] = {
+    val projectName = "snowplow-stream-loader"
+    case class FileConfig(config: File = new File("."))
+    val parser:OptionParser[FileConfig] = new scopt.OptionParser[FileConfig](projectName) {
```
Comment (BenFradet) [inline core/src/main/scala/com.snowplowanalytics.stream/loader/StreamLoaderApp.scala]: missing space

File: core/src/main/scala/com.snowplowanalytics.stream/loader/StreamLoaderApp.scala [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+import pureconfig._
+
+// This project
+import sinks._
+import model._
+
+/**
+ * Main entry point for the Elasticsearch sink
+ */
+trait StreamLoaderApp extends App {
```
Comment (BenFradet) [inline core/src/main/scala/com.snowplowanalytics.stream/loader/StreamLoaderApp.scala]: could we move away from using `App`?

File: core/src/main/scala/com.snowplowanalytics.stream/loader/StreamLoaderApp.scala [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+        System.err.println(s"configuration error: $e")
+        System.exit(1)
+        None
+       case Right(c) => Some(c)
+    }
+
+    streamLoaderConf
+  }
+
+  val config:StreamLoaderConfig = parseConfig().get
```
Comment (BenFradet) [inline core/src/main/scala/com.snowplowanalytics.stream/loader/StreamLoaderApp.scala]: please don't use `get`

File: core/src/main/scala/com.snowplowanalytics.stream/loader/model.scala [hunk 13]
```diff
@@ -18,9 +18,7 @@
  */
 
 package com.snowplowanalytics.stream.loader
-
 import java.text.SimpleDateFormat
-
```
Comment (BenFradet) [inline core/src/main/scala/com.snowplowanalytics.stream/loader/model.scala]: can you leave those lines

File: elasticsearch/src/main/scala/com/snowplowanalytics/stream/loader/ElasticsearchStreamLoaderApp.scala [hunk 14]
```diff
@@ -0,0 +1,125 @@
...
+        config.elasticsearch.get.client.endpoint,
+        config.elasticsearch.get.client.port,
+        config.elasticsearch.get.client.ssl,
+        config.elasticsearch.get.aws.region,
+        config.elasticsearch.get.aws.signing,
+        config.elasticsearch.get.client.username,
+        config.elasticsearch.get.client.password,
+        config.elasticsearch.get.client.maxTimeout,
+        config.elasticsearch.get.client.maxRetries,
+        config.elasticsearch.get.cluster.documentType,
```
Comment (BenFradet) [inline elasticsearch/src/main/scala/com/snowplowanalytics/stream/loader/ElasticsearchStreamLoaderApp.scala]: `get` shouldn't be needed, you case should be `case Some(esConfig)` => esConfig.client.endpoint`

File: elasticsearch/src/main/scala/com/snowplowanalytics/stream/loader/clients/ElasticsearchBulkSender.scala [hunk 15]
```diff
@@ -0,0 +1,166 @@
...
+  private val restClientBuilder = RestClient.builder(formedHost)
+    .setHttpClientConfigCallback(httpClientConfigCallback.asInstanceOf[RestClientBuilder.HttpClientConfigCallback])
+
+  if (username.orNull != null && password.orNull != null) {
+    val userpass = BaseEncoding.base64().encode(s"${username.get}:${password.get}".getBytes(Charsets.UTF_8))
+    val headers:Array[Header] = Array(new BasicHeader("Authorization", s"Basic $userpass"))
+    restClientBuilder.setDefaultHeaders(headers)
+  }
+
+  private val client: HttpClient = HttpClient.fromRestClient(restClientBuilder.build())
```
Comment (BenFradet) [inline elasticsearch/src/main/scala/com/snowplowanalytics/stream/loader/clients/ElasticsearchBulkSender.scala]: why was the client instantiation changed?

File: core/src/main/scala/com.snowplowanalytics.stream/loader/StreamLoaderApp.scala [hunk 16]
```diff
@@ -0,0 +1,117 @@
...
+
+  val tracker = config.monitoring.map(e => SnowplowTracking.initializeTracker(e.snowplow))
+
+  val goodSink = config.sink.good match {
+    case "stdout" => Some(new StdouterrSink)
+    case "elasticsearch" => None
```
Comment (BenFradet) [inline core/src/main/scala/com.snowplowanalytics.stream/loader/StreamLoaderApp.scala]: This can't be part of the core since it's what you're trying to abstract over.

You can look at [stream-enrich](https://github.com/snowplow/snowplow/tree/master/3-enrich/stream-enrich) to know how we've abstracted over the streaming platform.

File: core/src/main/scala/com.snowplowanalytics.stream/loader/StreamLoaderApp.scala (hunk 11 above)
Comment (szareiangm) [reply core/src/main/scala/com.snowplowanalytics.stream/loader/StreamLoaderApp.scala]: This is from the current code. If I have done it in this PR, you would have added to it the too much work list :stuck_out_tongue_winking_eye: Also, as you know `App` has internal mechanisms that removing it, might not look straight-forward and with small change.

**Comment by BenFradet** [comment]
>I wish we would have a styling config file for this project.

Yup, that's a very good point, we've been slowly integrating scalafmt in other projects, I created #101.

>Transformers needed to be abstracted because they generate ElasticsearchObject. It could have been done in the es module

Agreed but they didn't need to be moved in this commit? If you look at the diff for this PR, almost half is related to transformers and that should just be a few lines. Do you see what I mean?

File: core/src/main/scala/com.snowplowanalytics.stream/loader/StreamLoaderApp.scala (hunk 12 above)
Comment (szareiangm) [reply core/src/main/scala/com.snowplowanalytics.stream/loader/StreamLoaderApp.scala]: The same used to be in the code. I just moved it. 
https://github.com/snowplow/snowplow-elasticsearch-loader/pull/100/files#diff-20d604dd7232c1b254b18a7fc7440f04L28

File: elasticsearch/src/main/scala/com/snowplowanalytics/stream/loader/clients/ElasticsearchBulkSender.scala (hunk 15 above)
Comment (szareiangm) [reply elasticsearch/src/main/scala/com/snowplowanalytics/stream/loader/clients/ElasticsearchBulkSender.scala]: a mistake! reverting this part.

File: core/src/main/scala/com.snowplowanalytics.stream/loader/StreamLoaderApp.scala (hunk 16 above)
Comment (szareiangm) [reply core/src/main/scala/com.snowplowanalytics.stream/loader/StreamLoaderApp.scala]: I will move it `es` module since core doesn't need for any reason yet. Having it like `stream-enrich` is a long shot for this ticket.

**Comment by szareiangm** [comment]
> Agreed but they didn't need to be moved in this commit? If you look at the diff for this PR, almost half is related to transformers and that should just be a few lines. Do you see what I mean?

You are right. I will more cautious about amount of changes in each PR for next time. I should have broken down this issue, but now is a bit late. 😞  I addressed your comments except two (`good sink` change and eliminating `App`), that I wanted to defer for next pull requests since they might involve too much work for this PR.

File: core/src/test/scala/com.snowplowanalytics.stream.loader/PlainJsonTransformerSpec.scala [hunk 17]
```diff
@@ -12,23 +12,35 @@
...
-      val elasticsearchObject =
-        result._2.getOrElse(throw new RuntimeException("Plain Json failed transformation"))
-      elasticsearchObject.getIndex must_== "snowplow"
-      elasticsearchObject.getType must_== "PlainJson"
-      elasticsearchObject.getSource must_== input
+    "successfully convert a  plain JSON to an JsonRecord" in {
+      val input = """{"key1":"value1","key2":"value2","key3":"value3"}"""
+
+      val result = new PlainJsonTransformer("snowplow")
+        .fromClass(input -> JsonRecord(parse(input).asInstanceOf[JObject], "snowplow").success)
```
Comment (BenFradet) [inline core/src/test/scala/com.snowplowanalytics.stream.loader/PlainJsonTransformerSpec.scala]: why is the `asInstanceOf` needed?

File: elasticsearch/src/test/scala/com/snowplowanalytics/stream/loader/clients/ElasticsearchBulkSenderSpec.scala [hunk 18]
```diff
@@ -10,54 +10,64 @@
...
   client.execute(createIndex(index)).await
 
-  "sendToElasticsearch" should {
+  "send" should {
     "successfully send stuff" in {
-      val data = List(("a", Success(new ElasticsearchObject(index, "t", "i", """{"s":"json"}"""))))
-      sender.sendToElasticsearch(data) must_== List.empty
+
+      val validInput: EmitterJsonInput = "good" -> JsonRecord(
+        parse("""{"s":"json"}""").asInstanceOf[JObject],
```
Comment (BenFradet) [inline elasticsearch/src/test/scala/com/snowplowanalytics/stream/loader/clients/ElasticsearchBulkSenderSpec.scala]: same

File: core/src/main/scala/com.snowplowanalytics.stream/loader/JsonRecord.scala [hunk 19]
```diff
@@ -18,10 +18,12 @@
...
+
 /**
  * Format in which Snowplow events are buffered
  *
  * @param json The JSON string for the event
- * @param id The event ID (if it exists)
+ * @param partition partition to send the data to, currently not used.
  */
-case class JsonRecord(json: String, id: Option[String])
+case class JsonRecord(json: JObject, partition: String)
```
Comment (BenFradet) [inline core/src/main/scala/com.snowplowanalytics.stream/loader/JsonRecord.scala]: can we move this out of this PR, this doesn't seem related to abstracting the core module

File: core/src/main/scala/com.snowplowanalytics.stream/loader/transformers/JsonTransformer.scala [hunk 20]
```diff
@@ -41,68 +38,53 @@ import com.snowplowanalytics.snowplow.analytics.scalasdk.json.EventTransformer._
...
  * @param documentIndex the elasticsearch index name
- * @param documentType the elasticsearch index type
  */
-class SnowplowElasticsearchTransformer(documentIndex: String, documentType: String)
-    extends ITransformer[ValidatedRecord, EmitterInput]
+class JsonTransformer(documentIndex: String)
+    extends ITransformer[ValidatedJsonRecord, EmitterJsonInput]
     with StdinTransformer {
 
+  val log = LoggerFactory.getLogger(getClass)
```
Comment (BenFradet) [inline core/src/main/scala/com.snowplowanalytics.stream/loader/transformers/JsonTransformer.scala]: is this used?

File: core/src/main/scala/com.snowplowanalytics.stream/loader/transformers/JsonTransformer.scala [hunk 21]
```diff
@@ -41,68 +38,53 @@ import com.snowplowanalytics.snowplow.analytics.scalasdk.json.EventTransformer._
  * Class to convert successfully enriched events to EmitterInputs
  *
  * @param documentIndex the elasticsearch index name
- * @param documentType the elasticsearch index type
  */
-class SnowplowElasticsearchTransformer(documentIndex: String, documentType: String)
-    extends ITransformer[ValidatedRecord, EmitterInput]
+class JsonTransformer(documentIndex: String)
```
Comment (BenFradet) [inline core/src/main/scala/com.snowplowanalytics.stream/loader/transformers/JsonTransformer.scala]: I think the rename is confusing wrt the `PlainJsonTransformer`, maybe `EnrichedEventJsonTransformer` wdyt?

File: core/src/main/scala/com.snowplowanalytics.stream/loader/transformers/BadEventTransformer.scala [hunk 22]
```diff
@@ -35,11 +39,10 @@ import Scalaz._
 /**
  * Class to convert bad events to ElasticsearchObjects
  *
- * @param the elasticsearch index name
- * @param the elasticsearch index type
+ * @param documentIndex elasticsearch index name
  */
-class BadEventTransformer(documentIndex: String, documentType: String)
```
Comment (BenFradet) [inline core/src/main/scala/com.snowplowanalytics.stream/loader/transformers/BadEventTransformer.scala]: ideally, the `documentType` removal should have been in its own commit

File: core/src/main/scala/com.snowplowanalytics.stream/loader/transformers/BadEventTransformer.scala [hunk 23]
```diff
@@ -59,14 +64,15 @@ class BadEventTransformer(documentIndex: String, documentType: String)
    * @param record JsonRecord containing a bad event JSON
    * @return An ElasticsearchObject
    */
-  override def fromClass(record: ValidatedRecord): EmitterInput =
-    (record._1, record._2.map(j => new ElasticsearchObject(documentIndex, documentType, j.json)))
+  override def fromClass(record: ValidatedJsonRecord): EmitterJsonInput =
+    (record._1, record._2.map(j => j))
```
Comment (BenFradet) [inline core/src/main/scala/com.snowplowanalytics.stream/loader/transformers/BadEventTransformer.scala]: why is the map needed?

File: core/src/main/scala/com.snowplowanalytics.stream/loader/model.scala [hunk 24]
```diff
@@ -74,10 +74,11 @@ package model {
     username: Option[String],
     password: Option[String],
     maxTimeout: Long,
+    maxRetries: Int,
```
Comment (BenFradet) [inline core/src/main/scala/com.snowplowanalytics.stream/loader/model.scala]: this should have also been in its own commit

File: core/src/test/scala/com.snowplowanalytics.stream.loader/BadEventTransformerSpec.scala [hunk 25]
```diff
@@ -28,13 +35,11 @@ class BadEventTransformerSpec extends Specification {
...
         """{"line":"failed","errors":["Record does not match Thrift SnowplowRawEvent schema"]}"""
-      val result = new BadEventTransformer("snowplow", "bad")
-        .fromClass(input -> JsonRecord(input, None).success)
-      val elasticsearchObject =
-        result._2.getOrElse(throw new RuntimeException("Bad event failed transformation"))
-      elasticsearchObject.getIndex must_== "snowplow"
-      elasticsearchObject.getType must_== "bad"
-      elasticsearchObject.getSource must_== input
+      val result = new BadEventTransformer("snowplow")
+        .fromClass(input -> JsonRecord(parse(input).asInstanceOf[JObject], null).success)
```
Comment (BenFradet) [inline core/src/test/scala/com.snowplowanalytics.stream.loader/BadEventTransformerSpec.scala]: same with the `asInstanceOf` please don't use them

File: core/src/main/scala/com.snowplowanalytics.stream/loader/transformers/PlainJsonTransformer.scala [hunk 26]
```diff
@@ -75,8 +73,11 @@ class PlainJsonTransformer(documentIndex: String, documentType: String)
    */
   private def toJsonRecord(jsonString: String): ValidationNel[String, JsonRecord] = {
     parseOpt(jsonString) match {
-      case Some(jvalue) => JsonRecord(jsonString, None).success
-      case None         => "Json parsing error".failureNel
+      case Some(jvalue) =>
+        JsonRecord(
+          jvalue.asInstanceOf[JObject] ~ ("id" -> UUID.randomUUID().toString),
```
Comment (BenFradet) [inline core/src/main/scala/com.snowplowanalytics.stream/loader/transformers/PlainJsonTransformer.scala]: asInstanceOf

File: elasticsearch/src/main/scala/com/snowplowanalytics/stream/loader/clients/ElasticsearchBulkSender.scala [hunk 27]
```diff
@@ -80,23 +80,28 @@ class ElasticsearchSenderHTTP(
     HttpClient.fromRestClient(restClientBuilder.build())
   }
 
-  implicit val strategy = Strategy.DefaultExecutorService
-
   // do not close the es client, otherwise it will fail when resharding
   override def close(): Unit = ()
 
-  override def sendToElasticsearch(records: List[EmitterInput]): List[EmitterInput] = {
+  def send(records: List[EmitterJsonInput]): List[EmitterJsonInput] = {
```
Comment (BenFradet) [inline elasticsearch/src/main/scala/com/snowplowanalytics/stream/loader/clients/ElasticsearchBulkSender.scala]: can you add an override keyword

File: core/src/test/scala/com.snowplowanalytics.stream.loader/PlainJsonTransformerSpec.scala (hunk 17 above)
Comment (szareiangm) [reply core/src/test/scala/com.snowplowanalytics.stream.loader/PlainJsonTransformerSpec.scala]: I found an example of converting JValue to JObject like this,  although I knew `asInstanceOf` is not your favorite. Let me check alternatives.

File: core/src/main/scala/com.snowplowanalytics.stream/loader/JsonRecord.scala (hunk 19 above)
Comment (szareiangm) [reply core/src/main/scala/com.snowplowanalytics.stream/loader/JsonRecord.scala]: removed it to have it with partition pr

File: core/src/main/scala/com.snowplowanalytics.stream/loader/model.scala (hunk 24 above)
Comment (szareiangm) [reply core/src/main/scala/com.snowplowanalytics.stream/loader/model.scala]: removed

File: core/src/main/scala/com.snowplowanalytics.stream/loader/transformers/BadEventTransformer.scala (hunk 22 above)
Comment (szareiangm) [reply core/src/main/scala/com.snowplowanalytics.stream/loader/transformers/BadEventTransformer.scala]: reverted.

File: core/src/main/scala/com.snowplowanalytics.stream/loader/transformers/BadEventTransformer.scala (hunk 23 above)
Comment (szareiangm) [reply core/src/main/scala/com.snowplowanalytics.stream/loader/transformers/BadEventTransformer.scala]: simplified.

File: core/src/main/scala/com.snowplowanalytics.stream/loader/transformers/JsonTransformer.scala (hunk 21 above)
Comment (szareiangm) [reply core/src/main/scala/com.snowplowanalytics.stream/loader/transformers/JsonTransformer.scala]: but all of them are `EnrichedEventJsonTransformer`

Comment (szareiangm) [reply core/src/main/scala/com.snowplowanalytics.stream/loader/transformers/JsonTransformer.scala]: renamed, although I am not fan of long names. you are the boss

File: core/src/main/scala/com.snowplowanalytics.stream/loader/transformers/PlainJsonTransformer.scala (hunk 26 above)
Comment (szareiangm) [reply core/src/main/scala/com.snowplowanalytics.stream/loader/transformers/PlainJsonTransformer.scala]: Json4s is a nice library but its inheritance kills! changed.

File: core/src/test/scala/com.snowplowanalytics.stream.loader/BadEventTransformerSpec.scala (hunk 25 above)
Comment (szareiangm) [reply core/src/test/scala/com.snowplowanalytics.stream.loader/BadEventTransformerSpec.scala]: removed.

File: elasticsearch/src/main/scala/com/snowplowanalytics/stream/loader/clients/ElasticsearchBulkSender.scala (hunk 27 above)
Comment (szareiangm) [reply elasticsearch/src/main/scala/com/snowplowanalytics/stream/loader/clients/ElasticsearchBulkSender.scala]: done

File: elasticsearch/src/test/scala/com/snowplowanalytics/stream/loader/clients/ElasticsearchBulkSenderSpec.scala (hunk 18 above)
Comment (szareiangm) [reply elasticsearch/src/test/scala/com/snowplowanalytics/stream/loader/clients/ElasticsearchBulkSenderSpec.scala]: removed

**Comment by szareiangm** [comment]
Addressed the comments. Another round of review if I may ask please? @BenFradet

**Comment by BenFradet** [comment]
sorry I couldn't give another review sooner, I'll have a look this afternoon :+1:

File: core/src/main/scala/com.snowplowanalytics.stream/loader/Emitter.scala [hunk 28]
```diff
@@ -0,0 +1,175 @@
...
+
+  /**
+   * Emits good records to Sink and bad records to Kinesis.
+   * All valid records in the buffer get sent to the sink in a bulk request.
+   * All invalid requests and all requests which failed transformation get sent to Kinesis.
+   *
+   * @param records List of records to send
+   * @return List of inputs which the sink rejected
+   */
+  def emit(records: SList[EmitterJsonInput]): SList[EmitterJsonInput] = {
```
[the remark this answers is not in the document]
Comment (szareiangm) [reply core/src/main/scala/com.snowplowanalytics.stream/loader/Emitter.scala]: done

File: core/src/main/scala/com.snowplowanalytics.stream/loader/Emitter.scala [hunk 29]
```diff
@@ -0,0 +1,175 @@
...
+  /**
+   * Splits the buffer into emittable chunks based on the
+   * buffer settings defined in the config
+   *
+   * @param records     The records to split
+   * @param byteLimit   emitter byte limit
+   * @param recordLimit emitter record limit
+   * @returns a list of buffers
+   */
+  def splitBuffer(
```
[the remark this answers is not in the document]
Comment (szareiangm) [reply core/src/main/scala/com.snowplowanalytics.stream/loader/Emitter.scala]: done

File: core/src/main/scala/com.snowplowanalytics.stream/loader/Emitter.scala [hunk 30]
```diff
@@ -0,0 +1,175 @@
...
+    // add any remaining items to the final buffer
+    if (curBuffer.nonEmpty) buffers += curBuffer.toList
+
+    buffers.toList
+  }
+
+  /**
+   * Closes the Sink client when the KinesisConnectorRecordProcessor is shut down
+   */
+  def shutdown(): Unit = bulkSender.close
```
[the remark this answers is not in the document]
Comment (szareiangm) [reply core/src/main/scala/com.snowplowanalytics.stream/loader/Emitter.scala]: done

File: core/src/main/scala/com.snowplowanalytics.stream/loader/Emitter.scala [hunk 31]
```diff
@@ -0,0 +1,175 @@
...
+
+  /**
+   * Handles records rejected by the JsonTransformer or by Sink
+   *
+   * @param records List of failed records
+   */
+  def fail(records: List[EmitterJsonInput]): Unit = {
```
[the remark this answers is not in the document]
Comment (szareiangm) [reply core/src/main/scala/com.snowplowanalytics.stream/loader/Emitter.scala]: override added.

File: core/src/main/scala/com.snowplowanalytics.stream/loader/transformers/EnrichedEventJsonTransformer.scala [hunk 32]
```diff
@@ -41,68 +37,52 @@ import com.snowplowanalytics.snowplow.analytics.scalasdk.json.EventTransformer._
...
+   * @return An EmitterJsonInput
    */
-  override def fromClass(record: ValidatedRecord): EmitterInput =
-    record.map(_.map(r =>
-      r.id match {
-        case Some(id) => new ElasticsearchObject(documentIndex, documentType, id, r.json)
-        case None     => new ElasticsearchObject(documentIndex, documentType, r.json)
-    }))
+  override def fromClass(record: ValidatedJsonRecord): EmitterJsonInput =
+    record.map(_.map(r => r))
```
[the remark this answers is not in the document]
Comment (szareiangm) [reply core/src/main/scala/com.snowplowanalytics.stream/loader/transformers/EnrichedEventJsonTransformer.scala]: replaced with `record`

File: core/src/main/scala/com.snowplowanalytics.stream/loader/transformers/PlainJsonTransformer.scala [hunk 33]
```diff
@@ -62,10 +62,8 @@ class PlainJsonTransformer(documentIndex: String, documentType: String)
    * @param record ValidatedRecord containing plain JSON
    * @return An EmitterInput
    */
-  override def fromClass(record: ValidatedRecord): EmitterInput = {
-    val uuid = UUID.randomUUID().toString // generate unique UUID for every Json document
-    record.map(_.map(r => new ElasticsearchObject(documentIndex, documentType, uuid, r.json)))
-  }
+  override def fromClass(record: ValidatedJsonRecord): EmitterJsonInput =
+    record.map(_.map(r => r))
```
[the remark this answers is not in the document]
Comment (szareiangm) [reply core/src/main/scala/com.snowplowanalytics.stream/loader/transformers/PlainJsonTransformer.scala]: replaced with `record`

File: core/src/test/scala/com.snowplowanalytics.stream.loader/EmitterSpec.scala [hunk 34]
```diff
@@ -40,46 +41,51 @@ import org.specs2.mutable.Specification
...
-        override def sendToElasticsearch(records: List[EmitterInput]): List[EmitterInput] =
-          List.empty
-        override def close(): Unit            = ()
-        override def logClusterHealth(): Unit = ()
-        override val tracker                  = None
+      val fakeSender = new BulkSender[EmitterJsonInput] {
+        def send(records: List[EmitterJsonInput]): List[EmitterJsonInput] = List.empty
+        def close(): Unit                                                 = ()
+        def logHealth(): Unit                                             = ()
+        val tracker                                                       = None
```
[the remark this answers is not in the document]
Comment (szareiangm) [reply core/src/test/scala/com.snowplowanalytics.stream.loader/EmitterSpec.scala]: added `override` to all

File: elasticsearch/src/main/scala/com/snowplowanalytics/stream/loader/ElasticsearchStreamLoaderApp.scala [hunk 35]
```diff
@@ -0,0 +1,145 @@
...
+              config.elasticsearch.get.cluster.index,
+              config.elasticsearch.get.cluster.documentType)
+          case PlainJson =>
+            new PlainJsonTransformer(
+              config.elasticsearch.get.cluster.index,
+              config.elasticsearch.get.cluster.documentType)
+          case Bad =>
+            new BadEventTransformer(
+              config.elasticsearch.get.cluster.index,
+              config.elasticsearch.get.cluster.documentType)
```
[the remark this answers is not in the document]
Comment (szareiangm) [reply elasticsearch/src/main/scala/com/snowplowanalytics/stream/loader/ElasticsearchStreamLoaderApp.scala]: I have moved it to `match`

File: core/src/main/scala/com.snowplowanalytics.stream/loader/Emitter.scala (hunk 28 above)
[the remark this answers is not in the document]
Comment (szareiangm) [reply core/src/main/scala/com.snowplowanalytics.stream/loader/Emitter.scala]: `NsqSourceExecutor` uses it

[the remark this answers is not in the document]
Comment (szareiangm) [reply core/src/main/scala/com.snowplowanalytics.stream/loader/Emitter.scala]: no. `NsqSourceExecutor` uses this one

**Comment by szareiangm** [comment]
addressed the comments @BenFradet . please check

**Comment by szareiangm** [comment]
Yay! Send me a Snowplow T-Shirt. I should wear it in the office 🙌 😋

**Comment by alexanderdean** [comment]
> Send me a Snowplow T-Shirt.

I've created an email thread with Nicki on our side to get that arranged!
